File: internal/api/activitypub/apimodel/nodeinfo.go
package apimodel

type NodeInfoPointerResponse struct {
	Links []NodeInfoPointerLink `json:"links"`
}

type NodeInfoPointerLink struct {
	Rel  string `json:"rel"`
	Href string `json:"href"`
}

type NodeInfoResponse struct {
	Version           string           `json:"version"`
	Software          NodeInfoSoftware `json:"software"`
	Protocols         []string         `json:"protocols"`
	Usage             NodeInfoUsage    `json:"usage"`
	OpenRegistrations bool             `json:"openRegistrations"`
	Metadata          map[string]any   `json:"metadata"`
}

type NodeInfoSoftware struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

type NodeInfoUsage struct {
	Users      NodeInfoUsers `json:"users"`
	LocalPosts int64         `json:"localPosts"`
}

type NodeInfoUsers struct {
	Total int64 `json:"total"`
}

File: ui/lib/api/nodeinfo.ts
import { getConfig } from "@/lib/config";

export interface NodeInfoResponse {
  version: string;
  software: NodeInfoSoftware;
  protocols: string[];
  usage: NodeInfoUsage;
  openRegistrations: boolean;
  metadata: Record<string, unknown>;
}

export interface NodeInfoSoftware {
  name: string;
  version: string;
}

export interface NodeInfoUsage {
  users: NodeInfoUsers;
  localPosts: number;
}

export interface NodeInfoUsers {
  total: number;
}

export async function getNodeInfo(): Promise<NodeInfoResponse> {
  const config = await getConfig();
  const response = await fetch(`${config.server_url}/nodeinfo/2.0`);
  if (!response.ok) {
    throw new Error("Failed to load server information");
  }
  return response.json() as Promise<NodeInfoResponse>;
}
